Adding per container stats for CRI runtimes

abhi opened the pull request:
For https://github.com/kubernetes/features/issues/552.

**What this PR does / why we need it**

This commit aims to collect per container log stats. The change was proposed as a part of #55905. The change includes change the log path from /var/pod/<pod uid>/containername_attempt.log to /var/pod/<pod uid>/containername/containername_attempt.log. The logs are collected by reusing volume package to collect metrics from the log path.
Fixes #55905

**Special notes for your reviewer:**
cc @Random-Liu

**Release note:**

```
Adding container log stats for CRI runtimes.
```

Random-Liu commented on pkg/kubelet/kuberuntime/helpers.go:
```diff
@@ -207,14 +207,19 @@ func getStableKey(pod *v1.Pod, container *v1.Container) string {
 // buildContainerLogsPath builds log path for container relative to pod logs directory.
 func buildContainerLogsPath(containerName string, restartCount int) string {
-	return fmt.Sprintf("%s_%d.log", containerName, restartCount)
+	return fmt.Sprintf("%s/%s_%d.log", containerName, containerName, restartCount)
```
I feel like we can use `restartCount` as log file name for now until we have more metadata to add. Just to make the path shorter.
@mrunalp @feiskyer WDYT?

abhi replied:
Done

Random-Liu commented on pkg/kubelet/kuberuntime/helpers.go:
```diff
@@ -207,14 +207,19 @@ func getStableKey(pod *v1.Pod, container *v1.Container) string {
 // buildContainerLogsPath builds log path for container relative to pod logs directory.
 func buildContainerLogsPath(containerName string, restartCount int) string {
-	return fmt.Sprintf("%s_%d.log", containerName, restartCount)
+	return fmt.Sprintf("%s/%s_%d.log", containerName, containerName, restartCount)
```
keep consistent `filepath.Join`

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -190,6 +190,11 @@ func (m *kubeGenericRuntimeManager) generateContainerConfig(container *v1.Contai
 	}
 
 	command, args := kubecontainer.ExpandContainerCommandAndArgs(container, opts.Envs)
+	logDir := fmt.Sprintf("%s/%s", buildPodLogsDirectory(pod.UID), container.Name)
```
Use your helper function. `BuildContainerLogsDirectory`

Random-Liu replied:
Remove this directory when remove container.

abhi replied:
good point. Will do

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_container_test.go:
```diff
@@ -61,7 +61,7 @@ func TestRemoveContainer(t *testing.T) {
 	assert.NoError(t, err)
 	// Verify container log is removed
-	expectedContainerLogPath := filepath.Join(podLogsRootDirectory, "12345678", "foo_0.log")
+	expectedContainerLogPath := filepath.Join(podLogsRootDirectory, "12345678", "/foo/foo_0.log")
```
Use `Join` instead of `/`.

"foo", "0.log"

abhi replied:
done

Random-Liu commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -423,3 +415,26 @@ func getCRICadvisorStats(ca cadvisor.Interface) (map[string]cadvisorapiv2.Contai
+	m := volume.NewMetricsDu(path)
+	logMetrics, err := m.GetMetrics()
+	if err != nil {
+		return nil
```
log the error.

abhi replied:
Done

Random-Liu commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -423,3 +415,26 @@ func getCRICadvisorStats(ca cadvisor.Interface) (map[string]cadvisorapiv2.Contai
 	return stats, nil
 }
+
+func getContainerLogStats(path string) *statsapi.FsStats {
```
Add `TODO`, cache this in container log manager.

abhi replied:
done

Random-Liu commented on pkg/kubelet/stats/cri_stats_provider_test.go:
```diff
@@ -416,3 +433,13 @@ func checkCRINetworkStats(assert *assert.Assertions, actual *statsapi.NetworkSta
+
+func makeFakeLogs(sandbox *critest.FakePodSandbox, containers ...string) {
+	for _, name := range containers {
+		os.MkdirAll(fmt.Sprintf("/var/log/pods/%s/%s/", types.UID(sandbox.GetMetadata().GetUid()), name), 755)
```
Use your helper function.

abhi replied:
done

Random-Liu commented on pkg/kubelet/stats/cri_stats_provider_test.go:
```diff
@@ -416,3 +433,13 @@ func checkCRINetworkStats(assert *assert.Assertions, actual *statsapi.NetworkSta
+}
+
+func deleteFakeLogs(sandbox *critest.FakePodSandbox) {
+	os.RemoveAll(fmt.Sprintf("/var/log/pods/%s", types.UID(sandbox.GetMetadata().GetUid())))
```
ditto.

abhi replied:
removed

Random-Liu commented on pkg/kubelet/stats/cri_stats_provider_test.go:
```diff
@@ -416,3 +433,13 @@ func checkCRINetworkStats(assert *assert.Assertions, actual *statsapi.NetworkSta
+
+func makeFakeLogs(sandbox *critest.FakePodSandbox, containers ...string) {
+	for _, name := range containers {
+		os.MkdirAll(fmt.Sprintf("/var/log/pods/%s/%s/", types.UID(sandbox.GetMetadata().GetUid()), name), 755)
```
`/var/log/pods` actual pod/container log directly. We should not create files there in unit test...

abhi replied:
yep went mock approach

Random-Liu commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -423,3 +415,26 @@ func getCRICadvisorStats(ca cadvisor.Interface) (map[string]cadvisorapiv2.Contai
 }
+
+func getContainerLogStats(path string) *statsapi.FsStats {
+	m := volume.NewMetricsDu(path)
```
`NewMetricsDu` returns an interface `MetricsProvider`. It should be easy to mock it.
Let's use a mock instead of using actual log stats.

abhi replied:
done

abhi commented:
@Random-Liu will update the PR and address the review comments. Thanks for reviewing

Random-Liu commented on pkg/kubelet/stats/cri_stats_provider_test.go:
```diff
@@ -385,17 +412,21 @@ func checkCRIRootfsStats(assert *assert.Assertions, actual statsapi.ContainerSta
-	assert.Equal(*rootFsInfo.Inodes, *actual.Logs.Inodes)
-	assert.Nil(actual.Logs.UsedBytes)
-	assert.Nil(actual.Logs.InodesUsed)
+func checkCRILogsStats(assert *assert.Assertions, actual statsapi.ContainerStats) {
```
Compare the value for `rootFsInfo` and  log stats. That's the point of the fake, right? :)

abhi replied:
done

Random-Liu commented on pkg/kubelet/stats/cri_stats_provider_test.go:
```diff
@@ -416,3 +452,10 @@ func checkCRINetworkStats(assert *assert.Assertions, actual *statsapi.NetworkSta
+
+func makeFakeLogStats(seed int) *volume.Metrics {
+	m := &volume.Metrics{}
+	m.Used = resource.NewQuantity(int64(0), resource.BinarySI)       //seed + offsetUsage),resource.BinarySI)
```
? comment not removed?

abhi replied:
done

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -190,6 +190,11 @@ func (m *kubeGenericRuntimeManager) generateContainerConfig(container *v1.Contai
 	}
 
 	command, args := kubecontainer.ExpandContainerCommandAndArgs(container, opts.Envs)
+	logDir := BuildContainerLogsDirectory(kubetypes.UID(pod.UID), container.Name)
```
Remove the directory in `removeContainerLog`

abhi replied:
done

abhi committed:
Adding per container stats for CRI runtimes

This commit aims to collect per container log stats. The
change was proposed as a part of #55905. The change includes
change of the log path from /var/pod/<pod uid>/containername_attempt.log
to /var/pod/<pod uid>/containername/containername_attempt.log.
The logs are collected by reusing volume package to collect
metrics from the log path.

Signed-off-by: abhi <abhi@docker.com>

abhi committed:
Test cases to verify container log stats

The commit contains test case modifications to test
and verify changes for container log stats feature.

Signed-off-by: abhi <abhi@docker.com>

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -840,8 +845,7 @@ func (m *kubeGenericRuntimeManager) removeContainerLog(containerID string) error
 	labeledInfo := getContainerInfoFromLabels(status.Labels)
-	annotatedInfo := getContainerInfoFromAnnotations(status.Annotations)
-	path := buildFullContainerLogsPath(labeledInfo.PodUID, labeledInfo.ContainerName, annotatedInfo.RestartCount)
+	path := BuildContainerLogsDirectory(labeledInfo.PodUID, labeledInfo.ContainerName)
```
Sorry... This is my fault... We can't remove the whole directory... There might be newer instance of container log with different restart count in this directory. Let's leave the container log directory to be deleted when pod deletion.

abhi replied:
yep :) Cool will update

Random-Liu commented:
/lgtm
/approve

feiskyer reviewed (approved):
While the change is not tied to any platform, volume.MetricsProvider doesn't support windows yet. Will open a new issue for windows.

LGTM

Random-Liu commented:
@andor44 Are you using cri-o?

andor44 commented:
No, Docker 1.12. I worded my original comment wrongly, by "has nothing to do with CRI" I meant we don't use a CRI runtime.

Random-Liu commented:
We made a change to avoid the backward compatibility issue, but that was in 1.11 https://github.com/kubernetes/kubernetes/pull/63268.

But when upgrading from 1.9 to 1.10, you will have this issue. I'll say this is expected, because:
1) We never officially support in-place Kubelet update, it is always recommended to drain the node first before upgrade.
2) CRI had a breaking change, and the log path change is one of the breaking changes. And the CRI version is also bumped up to v1alpha2 in 1.10. https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.10.md#before-upgrading

We can cherry-pick https://github.com/kubernetes/kubernetes/pull/63268 to 1.10 to fix this, because the fix is small. @yujuhong Are you ok with that?

But please note that in-place kubelet may break again in the future, because it is never supported. :)

yujuhong commented:
> We can cherry-pick #63268 to 1.10 for this, because the fix is small. @yujuhong Are you ok with that?

Sounds good.

feiskyer commented:
> We can cherry-pick #63268 to 1.10 for this, because the fix is small.

LGTM. Created #66096 for this.

andor44 commented:
@Random-Liu understandable, but it seems that point 1 is undocumented, or at least I can't find anything in the docs that would indicate that the kubelet does not support in-place upgrades. I know that a lack of warning does not indicate support, but lots of online guides and how-tos tell you to do just that. Given that other components deal with in-place upgrades just fine, it is not a big mental leap to assume kubelet does too. If the kubelet really doesn't support it and such breaks are to be expected it should be well documented IMO.

FWIW, the cluster in question has been upgraded from 1.5 all the way to 1.10 over the past 2 years or so, and this is the first time we ran into some serious incompatibility/issue.

* Issue asking this very question, no answer: https://github.com/kubernetes/kubernetes/issues/60556
* [Stackoverflow question](https://stackoverflow.com/questions/50558557/do-i-need-drain-the-node-to-upgrade-the-kubelet-version) asking about this quotes a [doc page](https://kubernetes.io/docs/getting-started-guides/ubuntu/upgrades/), but I can't find said quote on the page.
* CLI reference page, no mention: https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/
* CoreOS Tectonic docs, implies this should work: https://coreos.com/tectonic/docs/latest/reference/kubelet-upgrade.html
* Kubeadm docs, nothing: https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-upgrade/
* Random blogpost that's GKE specific, this is the only place that mentions draining: https://medium.com/retailmenot-engineering/zero-downtime-kubernetes-cluster-upgrades-aab4cac943d2

lucaim commented:
I have to say that I agree with @andor44 here. We probably made assumption based, at least partially, on the lack of documentation or the presence of the contrary and our experience related to that. I knew that on GKE for example the node is drained but this could have been for a cloud specific issue/best practice. For us, on bare metal, it worked all the times. I kindly suggest to explicitly put this in the documentation.

yujuhong commented:
@andor44 @lucaim I agree that the release note should call out issues that may lead to container restarts (or other problems among kubelet upgrade) if possible. This is our oversight and sorry for the trouble.

AFAIK, the kubelet in-place upgrade path is not tested, so these issues may not be caught and well-documented before release. Just like the stackoverflow question @andor44 mentioned, container restarts also happened in 1.7 to 1.8 upgrade as well. If you use pod manifest files on the node, they'd encounter upgrade issues even more frequently (arguably by design). Given the current situation, I'd recommend draining the node to avoid unexpected issues.

feiskyer commented:
Agreed. Since this particular issue has been mitigated, let's address those user-facing issues clearly in the future on both PR's changelog and `BeforeUpgrade` part.
